---
title: "AARDVARK: An Automated Reversion Detector for Variants Affecting Resistance Kinetics"
output: 
  rmarkdown::html_document:
   toc: true
   toc_depth: 3
vignette: >
  %\VignetteIndexEntry{AARDVARK: An Automated Reversion Detector for Variants Affecting Resistance Kinetics}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, echo=FALSE}
library(aardvark)
library(knitr)
suppressPackageStartupMessages( require( BSgenome.Hsapiens.UCSC.hg38, quietly=TRUE, warn.conflicts=FALSE ) )
```

David Quigley, PhD  
UCSF Department of Urology

## Introduction 

AARDVARK is an R package that identifies reversion mutations in DNA sequence data. 

Genetic variations that inactivate genes essential for homologous recombination repair of double strand breaks, most frequently *BRCA1* and *BRCA2*, drive susceptibility to several types of cancer. Cells with these variants are sensitized to platinum therapies such as cisplatin and PARP inhibitor therapies such as olaparib. Unfortunately, "reversion" mutations that re-activate these genes produce therapy resistance. Reversion mutations were first described in 2008 (Edwards Nature 2008 and Sakai Nature 2008). In a 2017 paper, [Quigley et al. Cancer Discovery 2017](https://pubmed.ncbi.nlm.nih.gov/28450426/), we demonstrated that reversion mutations can be detected in advanced prostate cancer by liquid biopsy.

Detecting reversion mutations by hand is tedious and requires expert curation by a bioinformatician who knows genetics. We wrote AARDARK to allow for automated detection of reversion mutations in aligned genome sequence data.

## A workflow for calling reversion mutations

1) Align and index raw sequence data from a solid or liquid tumor biopsy
2) Define one or more pathogenic variants that inactivate a gene of interest (e.g. a BRCA2 two nucleotide deletion)
3) Run AARDVARK to realign all of the reads within a window around the pathogenic variant(s) and interpret whether any reads support a reversion mutation
 
Command line scripts are provided to produce reports from standard file types (sequence data in BAM format and pathogenic variants in VCF format) without writing any code. 

## Example: manually define a read and pathogenic variant

In this first example we'll create read and variant objects manually; normally AARDVARK would create these object by scanning a BAM and VCF file.

First, we'll **define a pathogenic mutation** in the *BRCA2* gene. The *seq_ref* and *seq_alt* parameters are written as they would be in a VCF file. In this case a two nucleotide deletion of *TT* at chromosome 13 positions 32339658 and 32339659 is represented as a variant starting at position 32339657 changing from *CTT* to *C*.

The *transcript* parameter defines the transcript affected by that mutation. AARDVARK ships with a set of ensembl transcripts, and you can define your own.

```{r}
pathogenic_mut = aardvark::Mutation( chrom="chr13", 
                         pos=32339657, 
                         seq_ref = "CTT", 
                         seq_alt = "C", 
                         transcript=aardvark::transcript_BRCA2)
```

Next, we'll **define an alignment window** for the analysis. The window can be of any size.

```{r}
window_start = pathogenic_mut$pos - 3000
window_end =   pathogenic_mut$pos + 3000
AW = aardvark::AlignmentWindow( Hsapiens_version = BSgenome.Hsapiens.UCSC.hg38, 
                                chrom="chr13",
                                window_start = window_start,
                                window_end = window_end)
```

Now we'll **create a read object**. This example read was aligned by *bwa* to start with a 72 nucleotide soft clip, followed by a 79 nucleotide perfect match. The alignment *bwa* chose is encoded in a [CIGAR](https://en.wikipedia.org/wiki/Sequence_alignment#Representations) string.

```{r}
read_nt = "CAGCCTTAGCTTTTTACACAAGTTGTAGTAGAAAAACTTCTGTGAGTCAGACTTCATTACTTGAAGCAAAAAAAAGTTCCTTACACAAAGTTAAGGGAGTGTTAGAGGAATTTGATTTAATCAGAACTGAGCATAGTCTTCACTATTCACC"
read_original = aardvark::Read( qname="A00887:299:HWFYGDSXY:2:2674:25211:28682",
                     cigar = "72S79M",
                     chrom = "chr13",
                     pos = 32340564,
                     seq = DNAString( read_nt ),
                     qual = rep(37, 151) )
```

We can inspect the cigar ranges of the original read to show the alignment:

```{r}
kable( read_original$cigar_ranges )
```

We'll now **realign the read**. 

```{r}
read_realigned = aardvark::realign_read( read = read_original, 
                                         align_window = AW, 
                                         pathogenic_mut)
```

The corrected alignment has a large deletion flanked by two perfect matches. We will discover that this deletion spans the pathogenic mutation, removing it while leaving the *BRCA2* gene in frame.

```{r}
kable( read_realigned$cigar_ranges )
```

Finally, we can assess whether this read is predicted to produce a reversion:

```{r}
read_realigned = aardvark::assess_reversion(read = read_realigned,
                                            transcript = aardvark::transcript_BRCA2,
                                            pathogenic = pathogenic_mut,
                                            align_window = AW)

read_summary = aardvark::summarize_candidates( list( read_realigned ), 
                                               transcript=aardvark::transcript_BRCA2,
                                               pathogenic_mutation = pathogenic_mut)

kable( read_summary$summary )
```

The read is predicted to produce a reversion.

We can plot this reversion:

```{r, fig.width=8, fig.height=4}

library( BSgenome.Hsapiens.UCSC.hg38 )
library( biomaRt )
obj.BM =  useEnsembl(biomart = "ENSEMBL_MART_ENSEMBL", 
                         dataset = "hsapiens_gene_ensembl")
aardvark::plot_reversion_summary(read_summary$summary, 
                               genome_version = 38,
                               hsapiens_object = BSgenome.Hsapiens.UCSC.hg38,
                               biomart_object = obj.BM,
                               pos_start = 32339000,
                               pos_end = 32341952,
                               min_freq = 1,
                               text_size_adjustment=1)
                                 

```


## Assessments from AARDVARK

After a call to *assess_reversion*, the read object has a string value for evidence. This is AARDVARK's interpretation of the read's consequence. The possible values are:

* *read_not_informative*, meaning that AARDVARK cannot use this read to assess a reversion. Possible reasons would include it is too far away, in an intron region not predicted to affect splicing, or it could not be re-aligned. 
* *read_harbors_variant_in_excluded_region*, meaning that the read overlaps a region that marks it as unreliable. The main reason for this annotation would be the read overlaps a homopolymer region, which are a frequent source of sequencing errors on Illumina instruments. 
* *read_harbors_pathogenic_variant_no_reversion*, meaining that AARDVARK detected the pathogenic variant but did not identify any other alterations in the read that would produce a reversion. Knowing how many of these reads are observed is helpful for assessing the proportion of reverted to un-reverted reads.
* *read_harbors_splice_site_mutation_no_reversion*, meaning this read includes a predicted splice-acceptor or splice-donor mutation and isn't currently anticipated to produce a reversion. The splice mutation may be the pathogenic variant
* *reversion_read_includes_pathogenic_variant*, meaning the read includes both the original pathogenic mutation and a reversion alteration. These are the strongest evidence for a reversion, because the pathogenic mutation and the reversion mutation must have been on the same strand of DNA that was sequenced.
* *reversion_read_deletion_spans_pathogenic_variant*, meaning that the read is predicted to produce a reversion and includes a deletion that spans the pathogenic variant, producing an in-frame gene.
* *reversion_read_does_not_include_pathogenic_variant*, meaning the read is predicted to produce a reversion but the read does not cover the pathogenic mutation. One example of this would be a read that includes deletion that is downstream from a pathogenic deletion, and is predicted to produce a reversion. This form of evidence is less strong than *reversion_read_includes_pathogenic_variant*, since the evaluation relies on the inference that the mutation in the read came from the same DNA strand as the pathogenic mutation, but we can't formally show that with this read. Usually tumors lose the wild-type allele of a pathogenic allele by DNA copy loss.

