#' Plot a reversion map showing the physical locations of the reversions in summary sums,
#' 'reading in reversion data from a summary generated by AARDVARK
#'
#' @param filename AARDVARK summary file to read
#' @param genome_version either 38 or 19
#' @param hsapiens_object hsapiens object from either BSgenome.Hsapiens.UCSC.hg38 or BSgenome.Hsapiens.UCSC.hg19
#' @param biomart_object a biomart object matching the genome_version parameter, created with the biomaRt package
#' @param pos_start leftmost position for plot
#' @param pos_end rightmost position for plot
#' @param exclude_blacklist boolean, if TRUE don't plot reversions marked with the evidence "overlaps_blacklist"; defaults TRUE
#' @param min_freq minimum number of observations to plot a candidate; defaults 2
#' @export
plot_reversion_summary_from_file = function( filename,
                                   genome_version=38,
                                   hsapiens_object,
                                   biomart_object,
                                   pos_start, pos_end,
                                   exclude_blacklist=TRUE,
                                   min_freq=2){
    if( !file.exists( filename ) ){
        stop( paste("Could not open file", filename ) )
    }
    sr = read.table( filename, header=TRUE )
    sr = sr[sr$N >= min_freq,]
    if( exclude_blacklist ){
        sr = sr[sr$evidence != "read_harbors_variant_in_excluded_region",]
    }
    ord = order(sr$N, decreasing=FALSE)
    chrom = unique(sr$chrom)[1]
    if( genome_version==38 ){
        genome_version_names = c("GRCh38", "hg38")
        strack <- Gviz::SequenceTrack(hsapiens_object, chromosome = chrom)
    }else{
        genome_version_names = c("GRCh19", "hg19")
        strack <- Gviz::SequenceTrack(hsapiens_object, chromosome = chrom)
    }

    ensembl_transcript_id = unique( sr$transcript_id)
    reversions = unique( sr$reversion )
    feature_types = c("pathogenic")

    biomTrack <- Gviz::BiomartGeneRegionTrack(genome = genome_version_names[1],
                                        chromosome = chrom,
                                        start = pos_start, end = pos_end,
                                        filters=list("ensembl_transcript_id" = ensembl_transcript_id),
                                        transcriptAnnotation = "symbol",
                                        name = "exon", biomart = biomart_object)

    tokens = strsplit( sr$pathogenic_mutation[1], ":")[[1]]
    pos_start_pathogenic = as.numeric( tokens[3] )
    width_pathogenic = as.numeric( tokens[2] )-1
    if( tokens[1]=="D" ){
        path_id = tokens[2]
    }else{
        path_id = paste( tokens[1], tokens[2], collapse="", sep="")
    }
    path_track = Gviz::AnnotationTrack(start=pos_start_pathogenic, width=width_pathogenic,
                                 chromosome=chrom, strand="+",
                                 id=path_id,
                                 genome=  genome_version_names[2],
                                 name = "mut",
                                 group="pathogenic mut.")
    starts = c()
    widths = c()
    group_names = c()
    group_lengths = c()
    ids = c()
    for(i in 1:length(reversions)){
        segs = strsplit( reversions[i], "_" )[[1]]
        group_names = c( group_names, paste0("r",i, " ", sum( sr$reversion==reversions[i] ), "x" ) )
        group_lengths = c(group_lengths, length(segs))
        for( j in 1:length( segs ) ){
            tokens = strsplit( segs[j], ":")[[1]]
            widths = c( widths, as.numeric( tokens[2] )-1 )
            starts = c( starts, as.numeric( tokens[3] ) )
            if( tokens[1] == "D" ){
                ids = c(ids, tokens[2] )
            }else{
                ids = c(ids, paste( tokens[1], tokens[2], collapse="", sep=""))
            }
        }
    }

    aTrack <- Gviz::AnnotationTrack(start = starts, width = widths,
                              chromosome = chrom,
                              strand = rep("+", length(starts)),
                              id=ids,
                              group=rep(group_names, group_lengths ),
                              genome = genome_version_names[2],
                              name = "reversions")
    Gviz::feature(aTrack) = "reversions"
    Gviz::feature(path_track) = "pathogenic"
    Gviz::feature(biomTrack) = "gene"
    display_properties = list(background.title = "white",
                              col.title = "black",
                              col.axis="black")
    Gviz::displayPars(aTrack) <- display_properties
    Gviz::displayPars(biomTrack) = display_properties
    Gviz::displayPars(path_track) = display_properties

    gtrack = Gviz::GenomeAxisTrack()

    Gviz::plotTracks( list( strack, biomTrack, path_track, aTrack, gtrack ),
                from=pos_start, to = pos_end,
                shape = "box",
                pathogenic="cornflowerblue", reversions="black", gene="lightgrey",
                featureAnnotation = "id",
                groupAnnotation = "group",
                cex=0.5, collapse=FALSE)
}




#' Plot a reversion map showing the physical locations of the reversions in summary sums
#'
#' @param sums Summary of reversions
#' @param xlim x axis limits
#' @param height height of a single row; defaults 1
#' @param xinterval X axis interval between ticks in the label; defaults 250
#' @param exclude_blacklist boolean, if TRUE don't plot reversions marked with the evidence "overlaps_blacklist"; defaults TRUE
#' @param min_freq minimum number of observations to plot a candidate; defaults 2
plot_reversion_summary = function( sums, xlim, height=1,
                                   xinterval=250,
                                   exclude_blacklist=TRUE,
                                   min_freq=2){
    sr = sums$summary
    cigs = sums$cigar_ranges
    if( exclude_blacklist ){
        cigs = cigs[sr$evidence != "overlaps_blacklist"]
        sr = sr[sr$evidence != "overlaps_blacklist",]
    }
    cigs = cigs[ sr$N>=min_freq]
    sr = sr[ sr$N>=min_freq,]
    ord = order(sr$N, decreasing=FALSE)
    N_y = dim(sr)[1]
    plot( -1, -1, xlim=xlim,
          ylim=c( 0, N_y * height ),
          axes=FALSE,
          xlab="", ylab="", yaxs="i", xaxs="i")
    axis(1, seq( from=xlim[1], to=xlim[2], by=xinterval ), las=2 )

    cur_y = 1
    for(i in 1:N_y){
        idx = ord[i]
        if( i %% 2 == 0 ){
            rect( xlim[1], cur_y, xlim[2], (cur_y + height - 1), col="#eeeeee", border="#eeeeee")
        }
        for( rr in 1:dim( cigs[[idx]] )[1]){
            x1=cigs[[idx]]$ref_start[rr]
            x2=cigs[[idx]]$ref_end[rr]
            if( cigs[[idx]]$cigar_code[rr] == "D" ){
                rect( x1, cur_y, x2, (cur_y + height - 1), col="black", border="black")
            }else if( cigs[[idx]]$cigar_code[rr]  == "I" ){
                rect( x1, cur_y, x2, (cur_y + height - 1), col="#1f78b4",border="#1f78b4")
            }
        }
        cur_y = cur_y + height
    }
    box()
}

#' Barplot the frequency of each reversion within a sample
#'
#'
#' @param sums Summary of reversions
#' @param height height of each bar
#' @param xinterval interval of x axis ticks
#' @param exclude_blacklist exclude candidate reversions that intersect a blacklist region
#' @param min_freq minimum number of observations to plot a candidate, default 2
plot_reversion_frequency = function( sums, height=1, xinterval=10,
                                     exclude_blacklist=TRUE,
                                     min_freq=2){
    sr = sums$summary
    if( exclude_blacklist ){
        sr = sr[sr$evidence != "overlaps_blacklist",]
    }
    sr = sr[ sr$N>=min_freq,]
    ord = order(sr$N, decreasing=FALSE)
    N_y = dim(sr)[1]
    xlim=c(0, max(sr$N))
    plot( -1, -1, xlim=xlim,
          ylim=c( 0, N_y * height ),
          axes=FALSE,
          xlab="", ylab="", yaxs="i", xaxs="i")
    axis(1, seq( from=xlim[1], to=xlim[2], by=xinterval ) )
    cur_y = 1
    for(i in 1:N_y){
        idx = ord[i]
        rect( 0, cur_y, sr$N[ idx ], (cur_y + height - 1), col="black", border="black")
        cur_y = cur_y + height
    }
    box()
}
